subsys/net/ip/icmpv4.c

```diff
@@ -59,9 +70,346 @@ int net_icmpv4_finalize(struct net_pkt *pkt)
+	 * incremented by one.
+	 */
+	if (ptr >= opt_len) {
+		/* overflow count itself overflows, the original datagram
+		 * is considered to be in error and is discarded.
+		 */
+		if (overflow == 0x0F) {
+			goto drop;
+		}
+
+		/* No free entry to update Timestamp data */
+		if (net_pkt_write_u8(reply, ptr)) {
+			goto drop;
+		}
+
+		len++;
+
+		overflow++;
+		flag = (overflow << 4U) | flag;
+
+		if (net_pkt_write_u8(reply, flag)) {
+			goto drop;
+		}
+
+		len++;
+
+		if (net_pkt_write(reply, opt_data + offset, opt_len)) {
+			goto drop;
+		}
+
+		len += opt_len;
+
+		net_pkt_set_ipv4_opts_len(reply, len);
+
+		return 0;
+	}
+
+	switch (flag) {
+	case NET_IPV4_TS_OPT_TS_ONLY:
+		new_entry_len = sizeof(u32_t);
+		break;
+	case NET_IPV4_TS_OPT_TS_ADDR:
+		new_entry_len = addr_len + sizeof(u32_t);
+		break;
+	case NET_IPV4_TS_OPT_TS_PRES: /* TODO */
+	default:
+		goto drop;
+	}
+
+	/* So, there is a free entry to update Timestamp */
+	if (net_pkt_write_u8(reply, ptr + new_entry_len)) {
+		goto drop;
+	}
+
+	len++;
+
+	if (net_pkt_write_u8(reply, (overflow << 4) | flag)) {
+		goto drop;
+	}
+
+	len++;
+
+	skip = ptr - ptr_offset;
+	if (skip) {
+		/* Do not alter existed routes */
+		if (net_pkt_write(reply, opt_data + offset, skip)) {
+			goto drop;
+		}
+
+		len += skip;
+		offset += skip;
+	}
+
+	switch (flag) {
+	case NET_IPV4_TS_OPT_TS_ONLY:
+		if (net_pkt_write_be32(reply, htons(k_uptime_get_32()))) {
+			goto drop;
+		}
+
+		len += sizeof(u32_t);
+
+		offset += sizeof(u32_t);
+
+		break;
+	case NET_IPV4_TS_OPT_TS_ADDR:
+		if (net_pkt_write(reply, (void *)src, addr_len)) {
+			goto drop;
+		}
+
+		len += addr_len;
+
+		if (net_pkt_write_be32(reply, htons(k_uptime_get_32()))) {
+			goto drop;
+		}
+
+		len += sizeof(u32_t);
+
+		offset += (addr_len + sizeof(u32_t));
+
+		break;
+	}
+
+	if (opt_len > offset) {
+		if (net_pkt_write(reply, opt_data + offset, opt_len - offset)) {
+			goto drop;
+		}
+	}
+
+	len += opt_len - offset;
+
+	net_pkt_set_ipv4_opts_len(reply, len);
+
+	return 0;
+
+drop:
+	return -EINVAL;
+}
+
+static int icmpv4_reply_to_options(u8_t opt_type,
+				   u8_t *opt_data,
+				   u8_t opt_len,
+				   void *user_data)
+{
+	struct net_icmpv4_hdr_opts_data *ud =
+		(struct net_icmpv4_hdr_opts_data *)user_data;
+
+	if (opt_type == NET_IPV4_OPTS_RR) {
+		return icmpv4_update_record_route(opt_data, opt_len,
+						  ud->reply, ud->src);
+	} else if (opt_type == NET_IPV4_OPTS_TS) {
+		return icmpv4_update_time_stamp(opt_data, opt_len,
+						ud->reply, ud->src);
+	}
+
+	return 0;
+}
+
+static int icmpv4_handle_header_options(struct net_pkt *pkt,
+					struct net_pkt *reply,
+					const struct in_addr *src)
+{
+	struct net_icmpv4_hdr_opts_data ud;
+	u8_t len;
+
+	ud.reply = reply;
+	ud.src = src;
+
+	if (net_ipv4_parse_hdr_options(pkt, icmpv4_reply_to_options, &ud)) {
+		return -EINVAL;
+	}
+
+	len = net_pkt_ipv4_opts_len(reply);
+
+	/* IPv4 header should ends in 32 bit boundary */
```

Comment:
the comment talks about ipv4 hdr, but you are actually verifying the optional headers length here. 
ipv4 hdr is by nature always fitting in 32 bits boundary. So just change "IPv4 header" to "IPv4 optional header part" something like that.